### PR TITLE
Generalize RefTypes

### PR DESCRIPTION
**Pull request by odersky**
Based on #3061.

The aim of this PR is to eliminate the various subtypes of TermRef and TypeRef and replace them by generalizing Name Name/Signature and Symbol to a new concept of `Designator`. 

The expected benefits of doing this are:

 - Drop the kludgey way we deal with "shadowed names" by making all private references symbolic
 - Allow to eliminate TypeArgRefs

### CODE REVIEW

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
test performance please

**Comment by liufengyun** [comment]
@odersky I just checked the logs, compiling of dotty project succeeds, but compilation of the [scala-library](https://github.com/liufengyun/bench/tree/master/tests/scala-library) project throws exception.

**Comment by smarter** [comment]
@lihaoyi Oh, the scala-library in the benchmarks is not the same as the one from the submodule? I think that would make more sense.

**Comment by liufengyun** [comment]
@smarter Exactly, it's not the same scala-library in the Dotty repo. It's intentionally made to be not the same, as the Dotty scala-library is growing, and we want a subset of stdlib that could be compiled since last october.

**Comment by odersky** [comment]
test performance please

File: compiler/src/dotty/tools/dotc/core/Designator.scala [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package dotty.tools
+package dotc
+package core
+
+import Names._
+
+/** Defines a common superclass of Name and Symbol and its Term/Type variants
+ *  Designators are used in reference type to identity what is referred to.
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/Designator.scala]: typo: identity -> identify

File: docs/docs/reference/multiversal-equality.md [hunk 2]
```diff
@@ -98,5 +100,10 @@ The precise rules for equality checking are as follows.
     instance if the implicit search for `Eq[T, T]` succeeds
     and constructs an instance different from `eqAny`.
 
+    Here _lifting_ a type `S` means replacing all references to  abstract types
+    in covariant positions of `S` by their upper bound, and to replacing
+    all refinement types in in covariant positions of `S` by their parent.
+    The purpose if lifte
```
Comment (smarter) [inline docs/docs/reference/multiversal-equality.md]: Sentence wasn't finished.

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 3]
```diff
@@ -1921,7 +1919,10 @@ object Types {
      *  the public name.
      */
     def shadowed(implicit ctx: Context): NamedType =
-      NamedType(prefix, designatorName.derived(ShadowedName))
+      designator match {
+        case designator: Symbol => this
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/Types.scala]: Is this case ever reached? It doesn't really seem to make sense.

File: compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala [hunk 4]
```diff
@@ -166,16 +158,23 @@ class TreePickler(pickler: TastyPickler) {
...
       }
-      else if (isLocallyDefined(sym) && tpe.signature.eq(Signature.NotAMethod)) {
+      else if (isLocallyDefined(sym)) {
         writeByte(if (tpe.isType) TYPEREFsymbol else TERMREFsymbol)
         pickleSymRef(sym); pickleType(tpe.prefix)
-      } else {
+      }
+      else if (tpe.hasFixedSym) {
+        assert(tpe.symbol.isClass && tpe.symbol.is(Flags.Scala2x), tpe.symbol.showLocated)
+        writeByte(TYPEREF) // should be changed to a new entry that keeps track of prefix, symbol & owner
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala]: Is this still a TODO?

**Comment by odersky** [comment]
Withdrawn in favor of #3147

**Pull request merged**